### PR TITLE
lightning: let ignore columns be compatible with tidb backend (#27850)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #27850 to release-5.3
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/30064
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/30064:release-5.3-c68791566d55
```

---

### What problem does this PR solve?
Let lightning config `mydumper.ignore-columns` be compatible with tidb backend. 

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?
Let the kv packend generated by encodeLoop contains the column names with ignore columns filtered instead of the raw column names. 

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 1]
```diff
@@ -154,8 +163,13 @@ func (s *mysqlSuite) TestWriteRowsIgnoreOnDup(c *C) {
 	c.Assert(err, IsNil)
 	row, err := encoder.Encode(logger, []types.Datum{
 		types.NewIntDatum(1),
+<<<<<<< HEAD
 	}, 1, []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, -1}, "1.csv", 0)
 	c.Assert(err, IsNil)
+=======
+	}, 1, []int{0, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1}, "1.csv", 0)
+	require.NoError(t, err)
+>>>>>>> c68791566... lightning: let ignore columns be compatible with tidb backend (#27850)
```
Comment (kennytm) [inline br/pkg/lightning/backend/tidb/tidb_test.go:172]: 👀

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 2]
```diff
@@ -198,8 +212,13 @@ func (s *mysqlSuite) TestWriteRowsErrorOnDup(c *C) {
 	c.Assert(err, IsNil)
 	row, err := encoder.Encode(logger, []types.Datum{
 		types.NewIntDatum(1),
+<<<<<<< HEAD
 	}, 1, []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, -1}, "3.csv", 0)
 	c.Assert(err, IsNil)
+=======
+	}, 1, []int{0, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1}, "3.csv", 0)
+	require.NoError(t, err)
+>>>>>>> c68791566... lightning: let ignore columns be compatible with tidb backend (#27850)
```
Comment (kennytm) [inline br/pkg/lightning/backend/tidb/tidb_test.go:221]: "

File: br/pkg/lightning/backend/tidb/tidb_test.go [hunk 3]
```diff
@@ -447,36 +467,61 @@ func (s *mysqlSuite) TestWriteRowsErrorDowngrading(c *C) {
 	c.Assert(err, IsNil)
 	row, err := encoder.Encode(logger, []types.Datum{
 		types.NewIntDatum(1),
+<<<<<<< HEAD
 	}, 1, []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, -1}, "7.csv", 0)
 	c.Assert(err, IsNil)
+=======
+	}, 1, []int{0, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1}, "7.csv", 0)
+	require.NoError(t, err)
+>>>>>>> c68791566... lightning: let ignore columns be compatible with tidb backend (#27850)
```
Comment (kennytm) [inline br/pkg/lightning/backend/tidb/tidb_test.go:476]: "

**Review by kennytm — APPROVED** [review]

**Comment by gozssky** [comment]
>[2021-11-25T02:54:11.317Z] PANIC: restore_test.go:1451: chunkRestoreSuite.TestEncodeLoopColumnsMismatch
[2021-11-25T02:54:11.317Z] 
[2021-11-25T02:54:11.317Z] ... Panic: runtime error: index out of range [0] with length 0 (PC=0x1348FD8)

🤔

**Review by glorv — APPROVED** [review]

**Pull request merged**